File: CSharp/QuantGateSignalsAPI/Subscriptions/StrategySubscription.cs
using QuantGate.API.Signals.Events;
using QuantGate.API.Signals.Proto.Stealth;
using QuantGate.API.Signals.Utilities;

namespace QuantGate.API.Signals.Subscriptions;

internal class StrategySubscription(APIClient client, EventHandler<StrategyEventArgs> handler,
                                    string strategyID, string streamID, string symbol,
                                    bool receipt = false, uint throttleRate = 0, object reference = null) :
    SubscriptionBase<StrategyUpdate, StrategyEventArgs>(client, StrategyUpdate.Parser, handler,
         new ParsedDestination(SubscriptionType.Strategy, SubscriptionPath.None,
                               ParsedDestination.StreamIDForSymbol(streamID, symbol),
                               symbol, strategyID: strategyID).Destination,
                               receipt, throttleRate, reference), ISymbolSubscription
{
    private readonly string _symbol = symbol;
    private readonly DataStream _stream = APIClient.ToStream(streamID);
    private readonly string _strategyID = strategyID;

    string ISymbolSubscription.Symbol => _symbol;
    DataStream ISymbolSubscription.Stream => _stream;

    protected override StrategyEventArgs HandleUpdate(StrategyUpdate update, object processed)
    {
        return new StrategyEventArgs(
            ProtoTimeEncoder.TimestampSecondsToDate(update.Timestamp),
            _symbol, _stream,
            _strategyID,
            update.EntryProgress / 1000.0,
            update.ExitProgress / 1000.0,
            ConvertLevel(update.PerceptionLevel),
            (GaugeSignal)update.PerceptionSignal,
            ConvertLevel(update.CommitmentLevel),
            (GaugeSignal)update.CommitmentSignal,
            ConvertLevel(update.EquilibriumLevel),
            (GaugeSignal)update.EquilibriumSignal,
            ConvertLevel(update.SentimentLevel),
            (GaugeSignal)update.SentimentSignal,
            (StrategySignal)update.Signal);
    }

    protected override StrategyEventArgs WrapError(SubscriptionError error) =>
        new(DateTime.UtcNow, _symbol, _stream, _strategyID, 0, 0, null,
             0, null, 0, null, 0, null, 0, StrategySignal.None, error);

    /// <summary>
    /// Converts a level to a nullable double value.
    /// </summary>
    /// <param name="level">The level to convert.</param>
    /// <returns>The converted nullable double value.</returns>
    private static double? ConvertLevel(uint level) => level == 0 ? null : (double?)((int)level - 1001) / 1000.0;
}

File: PHP/QuantGateSignalsAPI/Subscriptions/StrategySubscription.php
<?php

    namespace QuantGate\API\Signals\Subscriptions;

    use \QuantGate\API\Signals\APIClient;
    use \QuantGate\API\Signals\Events\StrategyUpdate;
    use \QuantGate\API\Signals\Events\SignalType; 
    use \QuantGate\API\Signals\Utilities;

    /**
     * Used to subscribe to a strategy stream subscription and convert the received messages.
     */
    class StrategySubscription extends SubscriptionBase
    {
        /**
         * The strategy subscribed to. Example enum values: PPr4.0, BTr4.0, Crb.8.4.
         * @var string
         */
        private string $strategyId;
        /**
         * Symbol to get the Strategy update data for.
         * @var string
         */
        private string $symbol;
        /**
         * Stream ID associated with the stream the client is connected to (realtime, delay, demo).
         * @var string
         */
        private string $stream;

        /** 
         * Creates a new StrategySubscription instance.          
         * @param  int        $id            The (integer) identifier to associate with this subscription on the server's end.
         * @param  string     $strategyId    The strategy to subscribe to. Example enum values: PPr4.0, BTr4.0, Crb.8.4.
         * @param  string     $symbol        Symbol to get the Strategy update data for.
         * @param  string     $stream        Stream ID associated with the stream the client is connected to (realtime, delay, demo).
         * @param  int        $throttleRate  Rate to throttle messages at (in ms). Enter 0 for no throttling.
         * @param  APIClient  $client        Reference to the parent APIClient instance to send updates to.
         */
        function __construct(int $id, string $strategyId, string $symbol, string $stream, int $throttleRate, APIClient $client)
        {
            // Set the properties.
            $this->strategyId = $strategyId;
            $this->symbol = $symbol;
            $this->stream = $stream;

            // Create the target destination.
            $destination = $this->createDestination($strategyId, $symbol, $stream);

            // Initialize values in the parent class.
            parent::__construct($destination, $id, $throttleRate, $client);
        }

        /**
         * Handles new messages from the stream and converts to subscription updates.
         * @param   $body   The raw message received from the stream.
         * @return void
         */
        public function handleMessage($body)
        {
            // Convert the message to a (Protobuf) strategy update object.
            $update = new \Stealth\StrategyUpdate();
            $update->mergeFromString($body);

            // Convert the values into usable values.
            $updateTime = Utilities::timestampSecondsToDate($update->getTimestamp());
            $entryProgress = $update->getEntryProgress() / 1000.0;
            $exitProgress = $update->getExitProgress() / 1000.0;
            $perceptionLevel = $this->convertGaugeLevel($update->getPerceptionLevel());
            $commitmentLevel = $this->convertGaugeLevel($update->getCommitmentLevel());
            $equilibriumLevel = $this->convertGaugeLevel($update->getEquilibriumLevel());
            $sentimentLevel = $this->convertGaugeLevel($update->getSentimentLevel());
            $perceptionSignal = SignalType::getFromGaugeSignal($update->getPerceptionSignal());
            $commitmentSignal = SignalType::getFromGaugeSignal($update->getCommitmentSignal());
            $equilibriumSignal = SignalType::getFromGaugeSignal($update->getEquilibriumSignal());
            $sentimentSignal = SignalType::getFromGaugeSignal($update->getSentimentSignal());
            $signal = SignalType::getFromStrategySignal($update->getSignal());            

            // Create the update object.
            $result = new StrategyUpdate($updateTime, $this->strategyId, $this->symbol, $this->stream, 
                                         $entryProgress, $exitProgress, $perceptionSignal, $perceptionLevel, 
                                         $commitmentSignal, $commitmentLevel, $equilibriumSignal,
                                         $equilibriumLevel, $sentimentSignal, $sentimentLevel, $signal);

            // Send the results back to the APIClient class.
            $this->client->emit('strategyUpdated', [$result]);
        }

        /**
         * Converts a raw gauge level value from an integer to a nullable float.
         * @param   int $level  The raw gauge level to convert.
         * @return  null|float
         */
        private function convertGaugeLevel(int $level) : ?float
        {
            // If zero, the result should be null.
            if ($level === 0)
                return null;

            // Convert non-zero values to a level.
            return ($level - 1001) / 1000.0;
        }

        /**
         * Creates the destination string that identifies this strategy stream.
         * @param   string  $strategyID Strategy to subscribe to. Example enum values: PPr4.0, BTr4.0, Crb.8.4.
         * @param   string  $symbol     Symbol to get the Strategy update data for.
         * @param   string  $stream     Stream ID associated with the stream the client is connected to (realtime, delay, demo).
         * @return  string
         */
        public static function createDestination(string $strategyID, string $symbol, string $stream) : string
        {            
            return "/strategy/".$strategyID."/".Utilities::streamIdForSymbol($stream, $symbol)."/".$symbol;
        }
    }

?>
    